### ISSUE TITLE
Option to automatically create user account who authenticate with SSO

### ISSUE THREAD

**Issue by kueronya** [directus/directus#2262]
# Feature Request
There should be an option to automatically create user accounts if they authenticate with a SSO provider. 

## What problem does this feature solve?
Right now only existent users are whitelisted for SSO authentication. So i have to expose the create user endpoint to unauthenticated users to create a new one before i'm able to authenticate them with SSO

## How do you think this should be implemented?
If the option is set, a user should be automatically created if they successfully authenticate with the SSO provider after the redirect from the oauth consent. Since the password, first and lastname are currently required for creating a new user account the fields should either be optional or the user should be created with dummy values.

## Would you be willing to work on this?
Since i'm just working a few days with directus i don't feel expierenced enough to do it on my own.

**Comment by rijkvanzanten** [comment]
@kueronya that would mean that _anyone_ with an account on one of those services automatically has access to the system. Isn't that insecure? What's the real life use case here?

**Comment by kueronya** [comment]
@rijkvanzanten The idea is to let unknown users/customers who will never interact with directus directly quickly register with your site without requiring them to provide a username/password first. There should also be an option to put them into a predefined role on creation. immediately granting them access to defined endpoints of the api.

**Comment by benhaynes** [comment]
Seems similar to this one: https://github.com/directus/app/issues/390

The difference here is that there would be accountability since they still need to register.

**Comment by benhaynes** [comment]
I think this is a bit of an edge-case, BUT if added with a domain whitelist it would be an awesome enterprise-level feature. eg: anyone with a `*@directus.io` email can sign up. Useful for ORGs with many users so they don't all need invites.

**Comment by kueronya** [comment]
I think it depends if you want to use directus more as a content management system or rather as a data api. For my planned use case i'm not so much interested in the content management features but rather having a solid rest api with granular permission system. 
Registering users during an oauth flow would be a really nice addition for community driven pages where content is mainly provided by the users itself through the website instead of the directus portal. 

Stackoverflow for example let users register with google/facebook without providing username/password. Strapi in it's current state is automatically creating user accounts for every different provider used (local, google, fb...) That way you can at least aggregate those 'logins' into one user account with the email as identity key.

**Comment by rijkvanzanten** [comment]
> automatically creating user accounts for every different provider used (local, google, fb...) That way you can at least aggregate those 'logins' into one user account with the email as identity key.

Directus uses the email of the existing account as the identity key for the external providers. EG if you have an account with email `admin@example.com` in Directus _and_ Google, you can use Google to login to Directus.

**Comment by kueronya** [comment]
> > automatically creating user accounts for every different provider used (local, google, fb...) That way you can at least aggregate those 'logins' into one user account with the email as identity key.
> 
> Directus uses the email of the existing account as the identity key for the external providers. EG if you have an account with email `admin@example.com` in Directus _and_ Google, you can use Google to login to Directus.

Yeah, i know, and that's really awesome. Just missing the possibility to register users during a sso without creating a local account in advance.

**Comment by benhaynes** [comment]
I'm not against adding an option for external users to register themselves as Directus (off by default). And I think that SSO is a good place to start for this since it offers a bit more accountability.

However, this is not something our core team would have time to work on in the near future. If someone wants to take this one on we can certainly supervise/help, but it would need to be a community effort.

**Comment by benhaynes** [comment]
@rijkvanzanten @WellingGuzman  — would you like to see this transferred to the App or API?

**Comment by WellingGuzman** [comment]
I believe if this could be an option it could be an extension, maybe a hook that allow you to handle non-existent users.

By default this might not work as expected, as Directus install two roles; Admin and Public, where should this new user go to? Which mean they will be either admin (probably not what anyone wants) or public (which means, not access to anything, basically the same thing of not being authenticated).

I go to, allow the user to do that, but not being a core feature or at least being OFF by default if we think this is a good idea to be a core feature.

The user can have the possibility to say:

- if the user doesn't exists and it's a successful SSO authentication, register them and add it into this role.
- If the user doesn't exists, it's a successful SSO authentication, and the email domain is in our whitelist let them in into this role.

Basically a custom authorization.



> @rijkvanzanten @WellingGuzman — would you like to see this transferred to the App or API?

It should be transfer to the API repository.
